Add configurable squadrons to AirstrikePower

penev92 opened the pull request:
This reworks `AirstrikePowerInfo` to allow for per-plane configuration of actor type, spawn offset, target offset ~and spawn delay~.
This way custom mods (because it probably doesn't have much space for that in the official ones) can do some pretty nice squadron formations and interesting airstrikes.
A follow-up PR can also add a configurable exit offset.

I consider this to be the first PR of 3 on the path to getting the D2k Ornithopter airstrike to match the original, with #17827 becoming the second PR in the series and a third PR using the logic from #17834.

Currently there are 3 major issues, 2 of which concern this PR and the 3rd concerning the follow-up PR.
 - ~With the new way planes are set up they are very hard to lint test. A proposed solution was to do the same thing we have for linting `HotkeyReferences`, but I consider that to be too ugly and out of place and also adds a lot more technical debt than it should, so I'm rather inclined to *not* lint test the actor references here.~    _**Fixed.**_
 - ~This probably needs an upgrade rule but I'm not 100% sure it can be completely automated. In case I'm wrong I'm going to need help with that.~    _**Fixed.**_
 - ~On the 3rd day of testing planes suddenly started turning nondeterministically to the left or right when configured for more than 1 pass. That is a big issue for when #17827 comes and I believe I tracked it down to something that sounds like a rounding issue to me. [Discord link to my finding.](https://discordapp.com/channels/153649279762694144/388282819371204608/704082376594030612)~    _**This was also discussed [on Discord](https://discordapp.com/channels/153649279762694144/388282819371204608/708076428536709190) and the planned fix is for a follow-up PR to add a specific trait/activity to handle the complex pathing.**_

penev92 committed:
Add configurable squadrons to AirstrikePower

MlemandPurrs commented:
Since we are touching AirstrikePower here: It would be nice if Airstrike can be made capable to attack air targets too. (see. Interceptors in CA mod)

pchote commented:
Please cherry pick 6fddc4df78ed4d8affa3ac7d3a3a748913f91fbb onto this branch, and then add
```csharp
[ActorReference(typeof(AircraftInfo))]
public IEnumerable<string> LintSquadActors { get { return Squad.Select(s => s.UnitType); } }
```
to the `AirstrikePowerInfo` to fix actor linting.

pchote commented on OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs:
```diff
@@ -142,47 +162,50 @@ public void SendAirstrike(Actor self, WPos target, bool randomize = true, int at
+					PlayLaunchSounds();
 
-					var a = w.CreateActor(info.UnitType, new TypeDictionary
+					var spawnOffset = squadMember.SpawnOffset.Rotate(attackRotation);
```
Please keep the comment
`// Includes the 90 degree rotation between body and world coordinates`
and the hardcoded 90 degree rotation here - it makes much more sense for the spawn and target offsets to be defined in body coordinates than world.

penev92 replied:
Added back.

pchote commented:
This should work for your update rule, after fixing my comment above:
```csharp
foreach (var airstrike in actorNode.ChildrenMatching("AirstrikePower"))
{
	var typeNode = airstrike.LastChildMatching("UnitType");
	var countNode = airstrike.LastChildMatching("SquadSize");
	var offsetNode = airstrike.LastChildMatching("SquadOffset");
	var unitType = typeNode != null ? typeNode.NodeValue<string>() : "badr.bomber";
	var squadSize = countNode != null ? countNode.NodeValue<int>() : 1;
	var offset = offsetNode != null ? offsetNode.NodeValue<WVec>() : new WVec(-1536, 1536, 0);

	var squadNode = new MiniYamlNode("Squad", "");
	for (var i = -squadSize / 2; i <= squadSize / 2; i++)
	{
		// Even-sized squads skip the lead plane
		if (i == 0 && (squadSize & 1) == 0)
			continue;

		var squadMemberNode = new MiniYamlNode(unitType, "");
		if (i != 0)
		{
			squadMemberNode.AddNode(new MiniYamlNode("SpawnOffset", FieldSaver.FormatValue(new WVec(i * offset.X, Math.Abs(i) * offset.Y, 0))));
			squadMemberNode.AddNode(new MiniYamlNode("TargetOffset", FieldSaver.FormatValue(new WVec(0, i * offset.Y, 0))));
		}

		squadNode.AddNode(squadMemberNode);
	}

	airstrike.AddNode(squadNode);
	airstrike.RemoveNode(typeNode);
	airstrike.RemoveNode(countNode);
	airstrike.RemoveNode(offsetNode);
}
```

penev92 committed:
Add support for [ActorReference] on properties

penev92 commented:
With a bunch of help from @pchote I think this PR is now ready for review.
I wanted to add an explanation of the [FLH](https://discordapp.com/channels/153649279762694144/388282819371204608/714590422860824576)/[local coordinate system](https://discordapp.com/channels/153649279762694144/388282819371204608/714590082903834644), but in the end I couldn't think of a good enough spot.

abcdefg30 commented:
Needs a rebase.

penev92 commented:
Rebased and fixed some issues. Should now be good to go.

This PR should not change any in-game behaviour in the official mods *except* in a couple of TD missions, where I have left comments.

penev92 commented:
Rebased.

Smittytron commented on mods/cnc/maps/nod07c/rules.yaml:
```diff
@@ -116,6 +116,9 @@ TRAN:
 		Prerequisites: gdi
+		Squad@override:
+			SquadMember:
+				UnitType: a10 # Assuming we want consistency with Nod07a and Nod07b. This should be discussed during review.
```
There is no script to send airstrikes in this mission and there is a fail trigger for killing or capturing the comm center, so this override isn't even needed. Nod07a and 07b are different missions, so no worries about consistency.

Smittytron commented:
> Testing map: 05c: Repair the GDI base
> OpenRA.Utility(1,1): Error: rules.yaml:173 refers to a trait field `SquadSize` that does not exist on `AirstrikePower`.
> 
> Testing map: 08b: Protect Mobius In Bratislava
> OpenRA.Utility(1,1): Error: rules.yaml:163 refers to a trait field `SquadSize` that does not exist on `AirstrikePower`.
> OpenRA.Utility(1,1): Error: rules.yaml:164 refers to a trait field `SquadOffset` that does not exist on `AirstrikePower`.

These two maps were merged recently.

penev92 commented:
Rebased, removed the changes to nod07c as discussed and added to gdi05c and gdi08b.

Mailaender commented:
Needs a rebase.

penev92 commented:
Rebased. I see RA now only has one plane for the parabombs thanks to #19792, so updated accordingly as well.

RoosterDragon commented on OpenRA.Mods.Common/Lint/CheckActorReferences.cs:
```diff
@@ -39,7 +39,8 @@ void ILintServerMapPass.Run(Action<string> emitError, Action<string> emitWarning
 		{
 			var actualType = traitInfo.GetType();
-			foreach (var field in actualType.GetFields())
+			var members = actualType.GetFields().Union(actualType.GetProperties().Cast<MemberInfo>());
```
- Suggest `Concat` since no need for de-duping `Union`.
- I think it'd be nicer to declare `IEnumerable<MemberInfo> members` rather than using `.Cast<>` but up to you.

penev92 replied:
Removing the `.Cast<>` results in trying to concat a `FieldInfo[]` with a `PropertyInfo[]` and failing. I think I'm going to need the cast.

RoosterDragon commented on OpenRA.Mods.Common/Traits/LintAttributes.cs:
```diff
@@ -13,7 +13,7 @@
 namespace OpenRA.Mods.Common.Traits
 {
-	[AttributeUsage(AttributeTargets.Field)]
+	[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
```
Should we update `AttributeUsage` on all `*ReferenceAttribute`s? to allow both fields and properties since they're now supported? Might be worth declaring a constant and reusing it for all such attributes.

RoosterDragon replied:
Actually I see that this is not the case and we'd have to update various lint checks to support properties.

Which might be nice so you can use properties with all lint checks without finding sometimes that isn't allowed, but that's a job for a different PR really if we did decide on such a thing.

RoosterDragon commented on OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs:
```diff
@@ -19,12 +18,25 @@
-		public readonly int SquadSize = 1;
-		public readonly WVec SquadOffset = new WVec(-1536, 1536, 0);
+		[FieldLoader.LoadUsing("LoadSquad")]
+		[Desc("A list of aircraft in the squad. Each has configurable UnitType, SpawnOffset and TargetOffset.")]
```
- `nameof(LoadSquad)`
- Does `nameof(AirstrikePowerSquadMember.UnitType)`, etc work and do the right thing here?

penev92 replied:
It should, yes. But note that this was written 2 years ago when we were still on .NET 4.6.1 probably 😆 
Will update.

penev92 replied:
> Feature 'constant interpolated strings' is not available in C# 7.3. Please use language version 10.0 or greater. [OpenRA.Mods.Common]csharp(CS8370)

Well I guess we ***can't*** use `nameof` with `DescAttribute` yet.

penev92 commented:
Rebased because of `UpdatePath` and fixed what I could from the above comments.

Mailaender commented on mods/ra/maps/top-o-the-world/rules.yaml:
```diff
@@ -61,13 +61,15 @@ powerproxy.spyplane:
 		ArrowSequence: arrow
 		ClockSequence: clock
 		CircleSequence: circles
+		Squad@override:
```
What is this overriding?

penev92 replied:
That is a good question and I don't have an answer. Removed the suffix as it seems unnecessary.

Mailaender commented on OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs:
```diff
@@ -48,6 +63,23 @@ public class AirstrikePowerInfo : SupportPowerInfo
+			var ret = new List<AirstrikePowerSquadMember>();
+
+			// A less than great workaround for the YamlParser merging nodes and not supporting a way to remove nodes that are not "traits".
+			var squadNode = yaml.Nodes.FirstOrDefault(n => n.Key == "Squad@override") ?? yaml.Nodes.FirstOrDefault(n => n.Key == "Squad");
```
This is really bad and hard to understand. I think the workaround should not live here, but rather in the rules without hardcoded MiniYaml keys.

```salt
^HQ:
	AirstrikePower:
		PauseOnCondition: lowpower
		Prerequisites: ~techlevel.superweapons
		Icon: airstrike
		ChargeInterval: 7500
		QuantizedFacings: 8
		Name: Air Strike
		Description: Deploy an aerial napalm strike.\nBurns buildings and infantry along a line.
		EndChargeSpeechNotification: AirstrikeReady
		SelectTargetSpeechNotification: SelectTarget
		InsufficientPowerSpeechNotification: InsufficientPower
		IncomingSpeechNotification: EnemyPlanesApproaching
		DisplayBeacon: True
		BeaconPoster: airstrike
		BeaconPosterPalette: beaconposter
		DisplayRadarPing: True
		CameraActor: camera
		ArrowSequence: arrow
		ClockSequence: clock
		CircleSequence: circles
		UseDirectionalTarget: True
		DirectionArrowAnimation: airstrikedirection
		SupportPowerPaletteOrder: 10
```

```salt
HQ:
	Inherits: ^HQ
	AirstrikePower:
		Squad:
			SquadMember@0:
				UnitType: a10
				SpawnOffset: -1536,-1536,0
				TargetOffset: 0,-1536,0
			SquadMember@1:
				UnitType: a10
			SquadMember@2:
				UnitType: a10
				SpawnOffset: -1536,1536,0
				TargetOffset: 0,1536,0
```

penev92 replied:
I can agree, but that means messing with many missions' rules and even with the default skirmish rules.

Mailaender replied:
You could also make it 
```yaml
    AirstrikePower:
        UnitTypes: a10, a10, a10
        SpawnOffsets: -1536,-1536,0,  0,0,0,  0,-1536,0 
        TargetOffsets: 0,-1536,0,   0,0,0,    0,1536,0
```
I think literally everything is better than hardcoded node value names.

penev92 replied:
Like I said on Discord - no. See comment below for details.

penev92 commented:
I am putting the following excerpt from Discord here not to bitch and moan, but because I already said there that I'm not going to do that latest suggestion:
> I have been growing increasingly tired of, and am currently in no mood to keep doing things in a way I know is wrong suboptimal
Just because that's how we've been doing them for a decade
That doesn't mean I'm scrapping the PR entirely, but am currently not in the mood to throw out progress (even if it's for consistency with everything else)

That said, if all else fails I may be persuaded to do the first suggestion, with the inheritance restructuring.

As a primary candidate for a solution here, however, I would like to suggest what @abcdefg30 said on Discord (jokingly or not) - fixing the issue at the source - the YAML parser.
 - This line with the hacky node selection in the trait's Load function goes away
 - The mission rules change to:
```diff
airstrike.proxy:
	AirstrikePower:
-		Squad@override:
+		-Squad:
+		Squad:
			SquadMember:
				UnitType: a10
```
 - And we get the ability to remove YAML nodes on overrides.

If that's accepted, then I have a working prototype at [penev92/yamlOverrides](https://github.com/penev92/OpenRA/tree/yamlOverrides) (the third commit, really) and we can move this along.
P.S.: Actually  https://github.com/OpenRA/OpenRA/commit/fdfa2725a710b79faf3b829d1b8e16f0e1c0dcd7

penev92 commented:
I plan to use #19987 to resolve the above issue.

penev92 commented:
Closing until the prererquisite PR is done to clear up the PR queue.

penev92 committed:
Add support for [WeaponReference] and [VoiceSetReference] on properties

penev92 committed:
Ran update rule

And applied manual fixes because the update rule
- doesn't know what actor to use outside of RA
- doesn't know when to apply `Squad` removal for overrides.

penev92 commented:
Not even 3 years later and we're ready to reopen! 🥳 
Thanks to #21839 (and others before it) overriding collections is now a thing.

P.S.: Nothing about the implementation has really changed, but I'm removing all labels, including "Needs +2" because a long time has passed since then.
